packages/extension-app/src/background/index.ts

```diff
@@ -2,40 +2,106 @@
 // eslint-disable-next-line @typescript-eslint/camelcase
 import init, { start_client } from '../../generated/polkadot_cli';
 import ws from '../../generated/ws';
-import { AnyJSON } from '../types';
-
-async function start(): Promise<void> {
-  /* Load WASM */
-  console.log('Loading WASM');
-  await init('./polkadot_cli_bg.wasm');
-  console.log('Successfully loaded WASM');
-
-  /* Build our client. */
-  console.log('Starting client');
-  const client = start_client(ws());
-  console.log('Client started', JSON.stringify(client));
-
-  /* A) Use the client directly */
-  // FIXME: use B
-  // https://github.com/paritytech/substrate-light-ui/issues/52
-  client.rpcSubscribe('{"method":"chain_subscribeNewHead","params":[],"id":1,"jsonrpc":"2.0"}', (r: AnyJSON) =>
-    console.log('[client] New chain head: ' + r)
-  );
-  client
-    .rpcSend('{"method":"system_networkState","params":[],"id":1,"jsonrpc":"2.0"}')
-    .then((r: AnyJSON) => console.log('[client] Network state: ' + r));
-
-  /* B) Use WasmProviderLite */
-  // const wasmProviderLite = new WasmProviderLite(client);
-  // wasmProviderLite.send('system_networkState', []).then(r => {
-  //   console.log('[WasmProviderLite] system_networkState resolved with', r);
-  // });
-  // wasmProviderLite.subscribe('n/a', 'chain_subscribeNewHead', [], (err: any, r: any) =>
-  //   console.log('[WasmProviderLite] Subscription notification : new chain head: ', r)
-  // );
+import { AnyJSON, MessageTypes, TransportRequestMessage } from '../types';
+
+// listen to all messages on the extension port and handle appropriately
+extension.runtime.onConnect.addListener((port: any): void => {
+  console.log('extension.runtime.onconnect() -> ', port);
+  console.log('Port name: ', port.name);
+  port.onMessage.addListener((data: any): void => {
+    console.log('background extensino.runtime.onConnect listner s=> ',  data);
+    handler(data, port);
+  });
+  port.onMessageExternal.addListener((data: any): void => {
+    console.log('on message external inside runtime on connect callback');
+  });
+  port.onDisconnect.addListener((): void => console.log(`Disconnected from ${port.name}`));
+});
+
+function handler<TMessageType extends MessageTypes> (payload: TransportRequestMessage<TMessageType>, port: chrome.runtime.Port): void {
+  const sender: chrome.runtime.MessageSender | undefined = port.sender;
+  const { id, message, request } = payload;
+
+  console.log('*** background handler *** ');
+
+  console.log('sender -> ', sender);
+  console.log('message -> ', message);
+  console.log('request -> ', request);
+
+  switch(message) {
+    case 'rpc.send':
+      return wasmRunner.rpcProxySend(payload, port);
+    case 'rpc.sendSubscribe':
+      return wasmRunner.rpcProxySubscribe(payload, () => {console.log('TODO implement the rpc subscribe callback...')}, port);
+    default:
+      throw new Error(`Unable to handle message of type ${message}`);
+  }
+}
+
+class WasmRunner {
```

Comment:
I don't see a point of having a class here. I'd put all the functions in this class in `wasmRunner.ts`, and export them

Reply:
true